DEPLOY.md

```diff
@@ -0,0 +1,101 @@
+# Deployment
+
+This repository contains two services: the RAP Agent and the RAP
+Controller. Both are deployed automatically on merge to main.
+
+
+## Agent deployment
+
+This runs within the secure backends (one instance per backend) and is
+managed by the [backend-server](https://github.com/opensafely-core/backend-server/) repo.
+
+A timer checks for new builds of the Docker image and pulls them down,
+so there may be a short delay (ten minutes maybe?) between the CI
+workflows completing and the new image being deployed.
+
+
+## Controller deployment
+
+We run a single instance of the controller which all the agents connect
+to. This is deployed as a Dokku app and at the time of writing runs on
+`dokku4`. A CI workflow connects to the host and runs the dokku deploy
+process with the newly built image.
+
+Below are the steps used to configure the app initially (though the
+exact config should be regarded as illustrative and may well have
+changed by the time of reading).
+
+Assume we have a server correctly configured by the [sysadmin](https://github.com/bennettoxford/sysadmin)
+tooling. This will ensure that Dokku is installed and configured with the
+necessary plugins and that a `rap-controller` user has been created.
+
+First we create the app and do some basic configuration:
+```bash
+dokku apps:create rap-controller
+
+# Set up the domain
+dokku domains:set rap-controller controller.opensafely.org
+dokku letsencrypt:enable rap-controller
```

Comment:
does this need to happen after you've deployed a container for the first time? doesn't really matter as you've said this is illustrative & someone running this should be able to handle that if it came up.

Reply:
It could well be. These instructions obviously aren't actually what I did because I didn't do the initial setup. I can well imagine this doesn't work until you've got an app live, but I can also imagine it _does_ work given that the LetsEncrypt plugin has to take over at least some URLs in order to do the auth dance.

Would be good to confirm one way or the other, but not urgent enough for me to do that now I think.

Reply:
ottmh if you haven't deployed an app at all, then port 80 isn't mapped to anything yet, so dokku attempts to redirect the url but letsencrypt finds a closed port & fails. doesn't need doing now.